### PR TITLE
Attach an address hash to PoC receipts

### PR DESCRIPTION
**Pull request by Vagabond**

### CODE REVIEW

File: src/transactions/v1/blockchain_txn_vars_v1.erl [hunk 1]
```diff
@@ -893,6 +893,8 @@ validate_var(?poc_max_hop_cells, Value) ->
     validate_int(Value, "poc_max_hop_cells", 100, 4000, false);
 validate_var(?poc_per_hop_max_witnesses, Value) ->
     validate_int(Value, "poc_per_hop_max_witnesses", 5, 50, false);
+validate_var(?poc_addr_hash_bytes, Value) ->
+    validate_int(Value, "poc_per_hop_max_witnesses", 4, 32, false);
```
Comment (abhay) [inline src/transactions/v1/blockchain_txn_vars_v1.erl:897]:
```suggestion
    validate_int(Value, "poc_addr_hash_bytes", 4, 32, false);
```

File: src/transactions/v1/blockchain_poc_receipt_v1.erl [hunk 2]
```diff
@@ -141,12 +153,24 @@ sign(Receipt, SigFun) ->
...
 
--spec is_valid(Receipt :: poc_receipt()) -> boolean().
-is_valid(Receipt=#blockchain_poc_receipt_v1_pb{gateway=Gateway, signature=Signature}) ->
-    PubKey = libp2p_crypto:bin_to_pubkey(Gateway),
-    BaseReceipt = Receipt#blockchain_poc_receipt_v1_pb{signature = <<>>},
-    EncodedReceipt = blockchain_txn_poc_receipts_v1_pb:encode_msg(BaseReceipt),
-    libp2p_crypto:verify(EncodedReceipt, Signature, PubKey).
+-spec is_valid(Receipt :: poc_receipt(), blockchain_ledger_v1:ledger()) -> boolean().
+is_valid(Receipt=#blockchain_poc_receipt_v1_pb{gateway=Gateway, signature=Signature, addr_hash=AH}, Ledger) ->
+    lager:info("checking hash ~p", [AH]),
```
Comment (evanmcc) [inline src/transactions/v1/blockchain_poc_receipt_v1.erl:158]: debugging

File: src/transactions/v1/blockchain_poc_receipt_v1.erl [hunk 3]
```diff
@@ -141,12 +153,24 @@ sign(Receipt, SigFun) ->
...
+    ValidHash = case blockchain_ledger_v1:config(?poc_addr_hash_bytes, Ledger) of
+                    {ok, Bytes} when is_integer(Bytes), Bytes > 0 ->
+                        AH == <<>> orelse (is_binary(AH) andalso byte_size(AH) == Bytes);
+                    _ ->
+                        AH == <<>>
```
Comment (vihu) [inline src/transactions/v1/blockchain_poc_receipt_v1.erl:163]: If the var is not set, should this be `AH == <<>> orelse AH == undefined` ?

Comment (Vagabond) [reply src/transactions/v1/blockchain_poc_receipt_v1.erl:163]: proto seems to always have it be <<>> if it's been serialized, but sure

File: src/ledger/v1/blockchain_ledger_v1.erl [hunk 4]
```diff
@@ -812,6 +813,20 @@ active_gateways(Ledger) ->
...
      ).
 
+-spec gateway_count(ledger()) -> non_neg_integer().
+gateway_count(Ledger) ->
+    AGwsCF = active_gateways_cf(Ledger),
+    cache_fold(
+      Ledger,
+      AGwsCF,
+      fun({_Address, _Binary}, Acc) ->
+              %% TODO perhaps we want to filter inactive GWs here?
```
Comment (vihu) [inline src/ledger/v1/blockchain_ledger_v1.erl:823]: We want to do this before landing the PR?

Comment (Vagabond) [reply src/ledger/v1/blockchain_ledger_v1.erl:823]: it's just used to size the bloom filter, so it being too big just gives us extra headroom on false positives.

**Review by andymck — APPROVED** [review]
lgtm

On the gateway fold count, I am surprised there is no easier way to get a count from rocks.  Even if it was just accessing internal properties where maybe some indicative count could be retrieved

File: include/blockchain_vars.hrl [hunk 5]
```diff
@@ -149,6 +149,8 @@
 %% during targeting
 -define(poc_witness_consideration_limit, poc_witness_consideration_limit).
 
+
+-define(poc_addr_hash_bytes, poc_addr_hash_bytes).
```
[the remark this answers is not in the document]
Comment (Vagabond) [reply include/blockchain_vars.hrl:153]: Good idea

**Pull request merged**